1-Draft/RFCNNNN-Automatic-Resource-Management.md

```diff
@@ -0,0 +1,197 @@
+Even an `IDisposable` check needs to be done only once.
+Moreover, we can even make a cache for all known types using our `TypeGen.ps1` utility.
+
+Another feature is that native objects can turn around in `PSObject` and the `Engine` should understand this too.
+
+### Annotating types / objects
+
+There are four options for annotating:
+
+- make a cache for all known types using our `TypeGen.ps1` utility
+- use ETS to enhance some types with this feature with default options
+
+    For example, if we define a TempFile type then users will not have to worry about deleting temporary files created by New-TemporaryFile cmdlet at all
+- embed this in the cmdlets so that they issue an already annotated object into the pipeline
+
+    For example, the same New-TemporaryFile or Open-Database cmdlet could do the annotation internally so that users have no need do this explicitly
+- explicitly annotate like `$tempFile.MakeDisposable(DisposableKind.AtRunspaceClose, { script to remove the temp file })`
+
+Note that there are many APIs that use disposable classes and now developers will have the opportunity to freely write objects of these types in the pipeline without fear of resource leakage.
+
+### Kinds of annotation
+
+Mentioned `DisposableKind` is based on context / scope and could be:
+
+- `Unknown`, it is default for new object. The `Engine` should annoutate the object. The default allow the `Engine` to process old code.
+- `Auto`, the `Engine` annoutates based on current context/scope
+- `AtGlobal`, dispose at default runspace close
+- `AtModuleUnload`
+- `AtRunspaceClose`
+- `AtCmdletClose/AtPipelineClose`
+- `AtScriptBlockClose`, dispose at current script block scope close
+
+Thus, a developer is always able to set a desired behavior.
+
+Moreover, we can always call MakeDisposable() in the script to adjust the behavior if the `Engine` cannot do this automatically.
+In this case, we will also need a method to force the mandatory code to execute like `$tempFile.PSDispose()`.
+It may also be required if we want to free an expensive resource before an end of the script block.
+
+### Implementtation details
+
+The main difficulty is that this is a generalized solution and users will expect it to work everywhere and as expected.
+Thus, it is possible the implementation will be adjusted based on feedback for a long time.
+
+From @lzybkr
+> I don't think PSVariable should implement IDisposable - but as an implementation detail, maybe SessionStateScope would.
+
+and
+> you need two things - better disposable support and a way to register arbitrary cleanup code - this would be analogous to C# using and finally statements.
+
+## Alternate Proposals and Considerations
+
+Add new `Dispose{}` script block to cmdlets.
+
+This approach:
+
+- only partially solves the problem of releasing up resources
+- completely assigns responsibility to an user implying his high qualifications
+- requires changes in Parser and tools
```

Comment:
It assigns responsibility to the author of the code. That may be the user, or it may be the author of a given module etc.

The same can be said for any C# library or code. Attempting to predict what users want instead seems like a problematic solution in my opinion. 🙂

Reply:
>  Attempting to predict what users want

If we were so pessimistic it wouldn’t be worth doing anything at all.